mpc/mpc-core/src/garble/label/mod.rs

```diff
@@ -82,289 +90,285 @@ impl From<[u8; 16]> for Delta {
+            .map(|value| Label::new(value))
+            .collect();
+
+        Self {
+            state: Full { delta },
+            labels: Arc::new(low),
         }
     }
 
-    /// Returns full labels from decoding information
-    pub fn from_decoding(
-        active_labels: Labels<G, state::Active>,
-        delta: Delta,
-        decoding: LabelsDecodingInfo<G>,
-    ) -> Result<Self, LabelError> {
-        Ok(Self {
-            group: active_labels.group,
-            state: state::Full::from_decoding(active_labels.state, delta, decoding.decoding)?,
-        })
-    }
+    /// Returns active labels corresponding to the `value`
+    pub fn select(&self, value: &Value) -> Result<ActiveLabels, EncodingError> {
+        if value.len() != self.labels.len() {
+            return Err(EncodingError::InvalidValue(self.len(), value.len()));
+        }
+
+        let active_labels = self
+            .labels
+            .iter()
+            .copied()
+            .zip(value.to_lsb0_bits().into_iter())
+            .map(|(low, level)| if level { low ^ self.get_delta() } else { low })
+            .collect::<Vec<_>>();
 
-    /// Returns active labels corresponding to a [`Value`]
-    pub fn select(&self, value: &Value) -> Result<Labels<G, state::Active>, LabelError> {
         Ok(Labels {
-            group: self.group.clone(),
-            state: self.state.select(value)?,
+            state: Active,
+            labels: Arc::new(active_labels),
         })
     }
 
-    /// Validates whether the provided active labels are authentic
-    pub fn validate(&self, labels: &Labels<G, state::Active>) -> Result<(), LabelError> {
-        for (pair, label) in self.state.iter().zip(labels.iter()) {
-            if !(label.value() == pair.low() || label.value() == pair.high()) {
-                return Err(LabelError::InauthenticLabels(labels.group.id().clone()));
-            }
+    pub(crate) fn from_decoding(
+        active: Labels<Active>,
+        delta: Delta,
+        decoding: Vec<bool>,
+    ) -> Result<Self, EncodingError> {
+        if active.labels.len() != decoding.len() {
+            return Err(EncodingError::InvalidDecodingLength(
+                active.labels.len(),
+                decoding.len(),
+            ));
         }
-        Ok(())
-    }
-
-    /// Returns wire labels
-    pub fn inner(&self) -> Vec<LabelPair> {
-        self.state.to_labels()
-    }
 
-    /// Returns labels as blocks
-    pub fn blocks(&self) -> Vec<[Block; 2]> {
-        self.inner()
-            .into_iter()
-            .map(|label| [label.low(), label.high()])
-            .collect()
+        Ok(Self {
+            state: Full { delta },
+            labels: Arc::new(
+                active
+                    .iter()
+                    .zip(decoding)
+                    .map(|(label, decoding)| {
+                        // If active label is logic high, flip it
+                        if label.permute_bit() ^ decoding {
```

Comment:
I put it on my TODO to add a label.decode(decoding) method to hide the .permute_bit() detail